lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -1092,9 +1096,223 @@ class DecomposeAtenMvOp : public OpRewritePattern<AtenMvOp> {
+//    X = B.reshape(*leading_dims, C, r*H, r*W)
+//
+// 'r' above is referred to as the 'upscale factor' or just 'factor' below.
+namespace {
+class DecomposeAtenPixelShuffleOp
+    : public OpRewritePattern<AtenPixelShuffleOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenPixelShuffleOp op,
+                                PatternRewriter &rewriter) const override {
+
+
+    Location loc = op.getLoc();
+    Value inValue = op.getSelf();
+    auto inType = inValue.getType().cast<BaseTensorType>();
+    auto maybeSizes = inType.getOptionalSizes();
+    if (!maybeSizes) {
+      return rewriter.notifyMatchFailure(
+          op, "Expected input tensor to have known rank.");
+    }
+    auto inShape = maybeSizes.value();
+    auto inRank = inShape.size();
+
+    // TODO: Handle dynamic shapes.
+    //
+    // Currently the decomposition of inputs with dynamic shapes results in a
+    // failure when lowering to linalg. Specifically, the error:
+    //
+    // "Unhandled case of expand/collapse. "
+    //
+    // is hit in the lowering of aten.view to the linalg dialect (see
+    // ConvertAtenViewOp in TorchToLinalg/DataMovement.cpp where this error is
+    // thrown).
+    //
+    // The issue is that there is no way for the lowering from torch to linalg
+    // to know that the view ops (reshape ops) created in this decomposition can
+    // be mapped directly to tensor.expand_shape and tensor_collapse_shape. We
+    // DO know that this is the case: the first reshape is an expand_shape
+    // and the second one is a collapse_shape. But we only know this because of
+    // the semantics of pixel_shuffle, by the time the lowering sees the view
+    // ops, this special context is lost.
+    //
+    // There are a few possible options to fixing the current approach, as I see
+    // it:
+    //
+    // 1) Lower pixel_shuffle to linalg directly. This would mean a bit of
+    // duplication of the logic for lowering from aten.permute to linalg, and
+    // would not be the most 'gradual' lowering approach. But perhaps the most
+    // straightforward and understable.
+    //
+    // 2) Rather than use aten.reshape ops in the decomposition below, insert
+    // tensor.expand_shape and tensor.collapse_shape ops directly. This would
+    // require inserting unrealized_conversion_cast ops to convert the types of
+    // the operands from !torch.tensor to to the MLIR built-in tensor type,
+    // which isn't ideal (unrealized_conversion_cast should only be used in
+    // conversion passes AFAIK (?)). UPDATE: not a viable option as we cannot
+    // make assumptions about the dialects that torch is lowered to.
+    //
+    // 3) Create 2 new ops in the torch dialect, with the same semantics as
+    // tensor.expand_shape and tensor.collapse_shape, but with !torch.tensor
+    // operands and results. Then use these ops in the decomposition below,
+    // instead of aten.reshape ops.
+    //
```

Comment:
Also, typically long TODO comments like this are best for github issues rather than inline in the code. Can you keep a short-form of `1)` as a TODO comment and move the rest to an issue? For example
```
// TODO: Lower pixel_shuffle to linalg directly. Pixel shuffle does a static reshape that is hard to recover through pure torch (view) ops, especially in dynamic cases.
```